Atomic/rpmwriter.py

```diff
@@ -0,0 +1,369 @@
+        return bytearray(struct.pack(">H", int(val)))
+
+    def _make_uint32(self, val):
+        return bytearray(struct.pack(">I", int(val)))
+
+    def _writebytearray(self, data):
+        self.written += len(data)
+        if not isinstance(data, bytearray):
+            data = bytearray(data)
+        self.out.write(data)
+
+    def _rpmlead(self):
+        def get_name(name, version, release):
+            name = "%s-%s-%s" % (name, version, release)
+            if len(name) > 65:
+                name = name[:65]
+            return (name + ("\0" * (66 - len(name)))).encode('ascii')
+        self._writebytearray(RpmWriter.MAGIC)
+        self._writebytearray(RpmWriter.MAJOR)
+        self._writebytearray(RpmWriter.MINOR)
+        self._writebytearray(RpmWriter.BINARY_TYPE)
+        self._writebytearray(RpmWriter.NOARCH)
+        self._writebytearray(get_name(self.name, self.version, self.release))
+        self._writebytearray(RpmWriter.OS)
+        self._writebytearray(RpmWriter.SIGNATURE_TYPE)
+        self._writebytearray(RpmWriter.RESERVED)
+
+    def pad(self, size):
+        while self.written % size != 0:
+            self._writebytearray([0])
+
+    def _signature(self, payload_size):
+        self._writebytearray(RpmWriter.SIGNATURE_MAGIC)
+        self._writebytearray(RpmWriter.SIGNATURE_VERSION)
+        self._writebytearray(RpmWriter.SIGNATURE_RESERVED)
+        self._writebytearray(self._make_uint32(1))
+        self._writebytearray(self._make_uint32(4))
+
+        self._writebytearray(self._make_uint32(RpmWriter.RPMSIGTAG_PAYLOADSIZE)) # sigtag_size
+        self._writebytearray(self._make_uint32(4)) # int32
+        self._writebytearray(self._make_uint32(0)) # offset
+        self._writebytearray(self._make_uint32(1)) # count
+
+        # payload
+        self._writebytearray(self._make_uint32(payload_size))
+
+        self.pad(8)
+
+    def _header(self):
+        header_section = bytearray()
+        store = bytearray()
+        self._writebytearray(RpmWriter.HEADER_MAGIC)
+        self._writebytearray(RpmWriter.HEADER_VERSION)
+        self._writebytearray(RpmWriter.HEADER_RESERVED)
+        for v in self.headers:
+            while (len(store) % v[4]) != 0:
+                store.append(0x0)
+            header_section += self._make_uint32(v[0]) # tag
+            header_section += self._make_uint32(v[1]) # type
+            header_section += self._make_uint32(len(store)) # offset
+            header_section += self._make_uint32(v[2]) # count
+            store += v[3] # value
+        self._writebytearray(self._make_uint32(len(self.headers)))
+        self._writebytearray(self._make_uint32(len(store)))
+        self._writebytearray(header_section)
+        self._writebytearray(store)
+
+    def _payload(self, out):
+        uncompressed_size = 0
+        def try_read(stdout, gzip_out):
+            written = 0
+            while True:
+                readable, _, _ = select.select([stdout], [], [], 0)
```

Comment:
Wow, is `select` needed here? I would assume that simple pipe with reads and writes is just enough.